sheepdog/transactions/upload/sub_entities.py

```diff
@@ -339,37 +343,55 @@ def _update_index(self):
                     document.patch()
             return
 
-        old_doc = document.to_json()
-        document.delete()
+        is_released = document.version is not None and document.metadata.get("release_number") is not None
 
-        # Re-create URL, since the filename might have changed.
-        # NOTE: We don't care about other URLs, since the file was not yet
-        # released, so it won't have any urls other than primary url
-        urls = [
-            generate_s3_url(
-                host=self._config['SUBMISSION']['host'],
-                bucket=self._config['SUBMISSION']['bucket'],
-                program=self.transaction.program,
-                project=self.transaction.project,
-                uuid=self.entity_id,
-                file_name=self.doc['file_name']
-            )
-        ]
-        urls_metadata = {
+        # if document is not released, allow recreation of urls and url_metadata
+        if not is_released:
+            document.urls = []
+            document.urls_metadata = {}
+
+        # generate new urls if none is specified
+        if not self.urls:
+            url = generate_s3_url(
+                    host=self._config['SUBMISSION']['host'],
+                    bucket=self._config['SUBMISSION']['bucket'],
+                    program=self.transaction.program,
+                    project=self.transaction.project,
+                    uuid=self.entity_id,
+                    file_name=self.doc['file_name']
+                )
+            self.urls = [url]
+
+        self.urls_metadata = {
             url: {
-                'state': 'registered', 'type': PRIMARY_URL_TYPE
-            } for url in urls
+                'state': 'registered',
+                'type': PRIMARY_URL_TYPE
+            } for url in self.urls
         }
 
+        # update urls and urls_metadata if it does not already exist
+        for url in self.urls:
+            if url not in document.urls:
+                document.urls.append(url)
+
+        # update urls metadata
+        for url, md in self.urls_metadata.items():
+            if url not in document.urls_metadata.keys():
+                document.urls_metadata[url] = md
+
+        old_doc = document.to_json()
+        document.delete()
+
         # Populate file metadata fields
         updated_fields = {
             'hashes': {'md5': self.doc['md5sum']},
             'size': self.doc['file_size'],
             'file_name': self.doc['file_name'],
-            'urls': urls,
-            'urls_metadata': urls_metadata,
-            'metadata': self.doc.get('metadata', {}),
-            'acl': self.node.acl or self.get_file_acl()
+            'urls': old_doc.get("urls"),
+            'urls_metadata': old_doc.get("urls_metadata"),
+            'acl': self.node.acl or self.get_file_acl(),
+            'version': old_doc.get("version"),
+            'metadata': old_doc.get('metadata', {}),
```

Comment:
so how do we prevent user from updating file_size and md5 for a validated file?

Reply:
for released file_size and md5 changes automatically trigger a new version. But I want to believe that can make these changes for unreleased files that are in a state that allows update